paparazzi2/src/main/java/app/cash/paparazzi/DeviceConfig.kt

```diff
@@ -0,0 +1,622 @@
+import com.android.ide.common.resources.configuration.LayoutDirectionQualifier
+import com.android.ide.common.resources.configuration.LocaleQualifier
+import com.android.ide.common.resources.configuration.NavigationMethodQualifier
+import com.android.ide.common.resources.configuration.NetworkCodeQualifier
+import com.android.ide.common.resources.configuration.NightModeQualifier
+import com.android.ide.common.resources.configuration.ScreenDimensionQualifier
+import com.android.ide.common.resources.configuration.ScreenOrientationQualifier
+import com.android.ide.common.resources.configuration.ScreenRatioQualifier
+import com.android.ide.common.resources.configuration.ScreenSizeQualifier
+import com.android.ide.common.resources.configuration.TextInputMethodQualifier
+import com.android.ide.common.resources.configuration.TouchScreenQualifier
+import com.android.ide.common.resources.configuration.UiModeQualifier
+import com.android.ide.common.resources.configuration.VersionQualifier
+import com.android.resources.Density
+import com.android.resources.Keyboard
+import com.android.resources.KeyboardState
+import com.android.resources.LayoutDirection
+import com.android.resources.Navigation
+import com.android.resources.NightMode
+import com.android.resources.ScreenOrientation
+import com.android.resources.ScreenRatio
+import com.android.resources.ScreenRound
+import com.android.resources.ScreenSize
+import com.android.resources.TouchScreen
+import com.android.resources.UiMode
+import com.google.android.collect.Maps
+import org.xmlpull.v1.XmlPullParser
+import org.xmlpull.v1.XmlPullParserException
+import org.xmlpull.v1.XmlPullParserFactory
+import java.io.File
+import java.io.FileInputStream
+import java.io.IOException
+import java.util.Properties
+
+/**
+ * Provides [FolderConfiguration] and [HardwareConfig] for various devices. Also provides utility
+ * methods to parse `build.prop` and `attrs.xml` to generate the appropriate maps.
+ *
+ * Defaults are for a Nexus 4 device.
+ */
+public data class DeviceConfig(
+  val screenHeight: Int = 1280,
+  val screenWidth: Int = 768,
+  val xdpi: Int = 320,
+  val ydpi: Int = 320,
+  val orientation: ScreenOrientation = ScreenOrientation.PORTRAIT,
+  val uiMode: UiMode = UiMode.NORMAL,
+  val nightMode: NightMode = NightMode.NOTNIGHT,
+  val density: Density = Density.XHIGH,
+  val fontScale: Float = 1f,
+  val layoutDirection: LayoutDirection = LayoutDirection.LTR,
+  val locale: String? = null,
+  val ratio: ScreenRatio = ScreenRatio.NOTLONG,
+  val size: ScreenSize = ScreenSize.NORMAL,
+  val keyboard: Keyboard = Keyboard.NOKEY,
+  val touchScreen: TouchScreen = TouchScreen.FINGER,
+  val keyboardState: KeyboardState = KeyboardState.SOFT,
+  val softButtons: Boolean = true,
+  val navigation: Navigation = Navigation.NONAV,
+  val screenRound: com.android.resources.ScreenRound? = null,
+  val released: String = "November 13, 2012"
+) {
+  val folderConfiguration: FolderConfiguration
```

Comment:
Off-topic: I’m surprised that this is called folder!

Reply:
Everyone loves resource qualifiers 🙃